File: Lawo.EmberPlusSharpTestVB/Model/TutorialTest.vb
''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''
' <copyright>Copyright 2012-2017 Lawo AG (http://www.lawo.com).</copyright>
' Distributed under the Boost Software License, Version 1.0.
' (See accompanying file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''

#Region "Using Declarations"
Imports System
Imports System.Linq
Imports System.Net.Sockets
Imports System.Threading
Imports System.Threading.Tasks
Imports Lawo.EmberPlusSharp.Model
Imports Lawo.EmberPlusSharp.S101
Imports Lawo.Threading.Tasks
'Imports Microsoft.TeamFoundation.Framework.Common
Imports Microsoft.VisualStudio.TestTools.UnitTesting
Imports Xunit.Sdk
#End Region

Public NotInheritable Class TutorialTestVB

    Private Sub New()
    End Sub

    Public Shared Sub DynamicConnectTest()
        Main()
    End Sub

    Public Shared Sub DynamicIterateTest()
#Region "Dynamic Iterate"
        Using cancelToken As New CancellationTokenSource()
            AsyncPump.Run(
            Async Function()
                Using client = Await ConnectAsync("localhost", 9000)
                    Using con = Await Consumer(Of MyRoot).CreateAsync(client)
                        WriteChildren(con.Root, 0)
                    End Using
                End Using
            End Function, cancelToken.Token)
        End Using
#End Region
    End Sub

    Public Shared Sub DynamicModifyTest()
#Region "Dynamic Modify"
        Using cancelToken As New CancellationTokenSource()
            AsyncPump.Run(
            Async Function()
                Using client = Await ConnectAsync("localhost", 9000)
                    Using con = Await Consumer(Of MyRoot).CreateAsync(client)
                        Dim root As INode = con.Root

                        ' Navigate to the parameters we're interested in.
                        Dim sapphire = DirectCast(root.Children.First(Function(c) c.Identifier = "Sapphire"), INode)
                        Dim sources = DirectCast(sapphire.Children.First(Function(c) c.Identifier = "Sources"), INode)
                        Dim fpgm1 = DirectCast(sources.Children.First(Function(c) c.Identifier = "FPGM 1"), INode)
                        Dim fader = DirectCast(fpgm1.Children.First(Function(c) c.Identifier = "Fader"), INode)
                        Dim dbValue = DirectCast(fader.Children.First(Function(c) c.Identifier = "dB Value"), IParameter)
                        Dim position = DirectCast(fader.Children.First(Function(c) c.Identifier = "Position"), IParameter)

                        ' Set parameters to the desired values.
                        dbValue.Value = -67.0
                        position.Value = 128L

                        ' We send the changes back to the provider with the call below. Here, this is necessary so that
                        ' the changes are sent before Dispose is called on the consumer. In a real-world application
                        ' however, SendAsync often does not need to be called explicitly because it is automatically
                        ' called every 100ms as long as there are pending changes. See AutoSendInterval for more
                        ' information.
                        Await con.SendAsync()
                    End Using
                End Using
            End Function, cancelToken.Token)
        End Using

#End Region
    End Sub

    Public Shared Sub ConnectionLostTest()
#Region "Connection Lost"
        Using cancelToken As New CancellationTokenSource()
            AsyncPump.Run(
            Async Function()
                Using client = Await ConnectAsync("localhost", 9000)
                    Using con = Await Consumer(Of MyRoot).CreateAsync(client)
                        Dim connectionLost = New TaskCompletionSource(Of Exception)()
                        AddHandler con.ConnectionLost, Sub(s, e) connectionLost.SetResult(e.Exception)

                        Console.WriteLine("Waiting for the provider to disconnect...")
                        Dim exception = Await connectionLost.Task
                        Console.WriteLine("Connection Lost!")
                        Console.WriteLine("Exception:{0}{1}", exception, Environment.NewLine)
                    End Using
                End Using
            End Function, cancelToken.Token)
        End Using
#End Region
    End Sub

    Public Shared Sub StaticIterateTest()
#Region "Static Iterate"
        Using cancelToken As New CancellationTokenSource()
            AsyncPump.Run(
            Async Function()
                Using client = Await ConnectAsync("localhost", 9000)
                    Using con = Await Consumer(Of SapphireRoot).CreateAsync(client)
                        WriteChildren(con.Root, 0)
                    End Using
                End Using
            End Function, cancelToken.Token)
        End Using
#End Region
    End Sub

    Public Shared Sub StaticReactToChangesTest()
#Region "Static React to Changes"
        Using cancelToken As New CancellationTokenSource()
            AsyncPump.Run(
            Async Function()
                Using client = Await ConnectAsync("localhost", 9000)
                    Using con = Await Consumer(Of SapphireRoot).CreateAsync(client)
                        Dim valueChanged = New TaskCompletionSource(Of String)()
                        Dim positionParameter = con.Root.Sapphire.Sources.Fpgm1.Fader.Position
                        AddHandler positionParameter.PropertyChanged, Sub(s, e) valueChanged.SetResult(DirectCast(s, IElement).GetPath())
                        Console.WriteLine("Waiting for the parameter to change...")
                        Console.WriteLine("A value of the element with the path {0} has been changed.", Await valueChanged.Task)
                    End Using
                End Using
            End Function, cancelToken.Token)
        End Using
#End Region
    End Sub

    Public Shared Sub StaticModifyTest()
#Region "Static Modify"
        Using cancelToken As New CancellationTokenSource()
            AsyncPump.Run(
            Async Function()
                Using client = Await ConnectAsync("localhost", 9000)
                    Using con = Await Consumer(Of SapphireRoot).CreateAsync(client)
                        Dim fader = con.Root.Sapphire.Sources.Fpgm1.Fader
                        fader.DBValue.Value = -67.0
                        fader.Position.Value = 128
                        Await con.SendAsync()
                    End Using
                End Using
            End Function, cancelToken.Token)
        End Using
#End Region
    End Sub

    Public Shared Sub CollectionNodeTest()
#Region "Collection Node"
        Using cancelToken As New CancellationTokenSource()
            AsyncPump.Run(
            Async Function()
                Using client = Await ConnectAsync("localhost", 9000)
                    Using con = Await Consumer(Of UnboundedSapphireRoot).CreateAsync(client)
                        For Each Source In con.Root.Sapphire.Sources.Children
                            Console.WriteLine(Source.Fader.Position.Value)
                        Next
                    End Using
                End Using
            End Function, cancelToken.Token)
        End Using
#End Region
    End Sub

    Public Shared Sub MixedIterateTest()
#Region "Mixed Iterate"
        Using cancelToken As New CancellationTokenSource()
            AsyncPump.Run(
            Async Function()
                Using client = Await ConnectAsync("localhost", 9000)
                    Using con = Await Consumer(Of MixedSapphireRoot).CreateAsync(client)
                        WriteChildren(con.Root, 0)
                    End Using
                End Using
            End Function, cancelToken.Token)
        End Using
#End Region
    End Sub

    Public Shared Sub MixedModifyTest()
#Region "Mixed Modify"
        Using cancelToken As New CancellationTokenSource()
            AsyncPump.Run(
            Async Function()
                Using client = Await ConnectAsync("localhost", 9000)
                    Using con = Await Consumer(Of MixedSapphireRoot).CreateAsync(client)
                        For Each Source In con.Root.Sapphire.Sources.Children
                            Source.Fader.DBValue.Value = -67.0
                            Source.Fader.Position.Value = 128
                            Source.Dsp.Input.LRMode.Value = LRMode.Mono
                            Source.Dsp.Input.Phase.Value = False
                        Next

                        Await con.SendAsync()
                    End Using
                End Using
            End Function, cancelToken.Token)
        End Using
#End Region
    End Sub

    ''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''

#Region "Main Method"
    Private Shared Sub Main()
        ' This is necessary so that we can execute async code in a console application.
        Using cancelToken As New CancellationTokenSource()
            AsyncPump.Run(
            Async Function()
                ' Establish S101 protocol
                Using client As S101Client = Await ConnectAsync("localhost", 9000)
                    ' Retrieve *all* elements in the provider database and store them in a local copy
                    Using con As Consumer(Of MyRoot) = Await Consumer(Of MyRoot).CreateAsync(client)
                        ' Get the root of the local database.
                        Dim root As INode = con.Root

                        ' For now just output the number of direct children under the root node.
                        Console.WriteLine(root.Children.Count)
                    End Using
                End Using
            End Function, cancelToken.Token)
        End Using
    End Sub
#End Region

#Region "S101 Connect Method"
    Private Shared Async Function ConnectAsync(host As String, port As Integer) As Task(Of S101Client)
        ' Create TCP connection
        Dim tcpClient = New TcpClient()
        Await tcpClient.ConnectAsync(host, port)

        ' Establish S101 protocol
        ' S101 provides message packaging, CRC integrity checks and a keep-alive mechanism.
        Dim stream = tcpClient.GetStream()
        Return New S101Client(tcpClient, AddressOf stream.ReadAsync, AddressOf stream.WriteAsync)
    End Function
#End Region

#Region "Write Children"
    Private Shared Sub WriteChildren(node As INode, depth As Integer)
        Dim indent = New String(" "c, 2 * depth)

        For Each child In node.Children
            Dim childNode = TryCast(child, INode)

            If childNode IsNot Nothing Then
                Console.WriteLine("{0}Node {1}", indent, child.Identifier)
                WriteChildren(childNode, depth + 1)
            Else
                Dim childParameter = TryCast(child, IParameter)

                If childParameter IsNot Nothing Then
                    Console.WriteLine("{0}Parameter {1}: {2}", indent, child.Identifier, childParameter.Value)
                End If
            End If
        Next
    End Sub
#End Region

#Region "Dynamic Root Class"
    ' Note that the most-derived subtype MyRoot needs to be passed to the generic base class.
    Private NotInheritable Class MyRoot
        Inherits DynamicRoot(Of MyRoot)
    End Class
#End Region

#Region "Static Database Types"
    Private NotInheritable Class SapphireRoot
        Inherits Root(Of SapphireRoot)

        Friend Property Sapphire As Sapphire
    End Class

    Private NotInheritable Class Sapphire
        Inherits FieldNode(Of Sapphire)

        Friend Property Sources As Sources
    End Class

    Private NotInheritable Class Sources
        Inherits FieldNode(Of Sources)

        <Element(Identifier:="FPGM 1")>
        Friend Property Fpgm1 As Source

        <Element(Identifier:="FPGM 2")>
        Friend Property Fpgm2 As Source
    End Class

    Private NotInheritable Class Source
        Inherits FieldNode(Of Source)

        Friend Property Fader As Fader

        <Element(Identifier:="DSP")>
        Friend Property Dsp As Dsp
    End Class

    Private NotInheritable Class Fader
        Inherits FieldNode(Of Fader)

        <Element(Identifier:="dB Value")>
        Friend Property DBValue As RealParameter

        Friend Property Position As IntegerParameter
    End Class

    Private NotInheritable Class Dsp
        Inherits FieldNode(Of Dsp)

        Friend Property Input As Input
    End Class

    Private NotInheritable Class Input
        Inherits FieldNode(Of Input)

        Friend Property Phase() As BooleanParameter

        <Element(Identifier:="LR Mode")>
        Friend Property LRMode As EnumParameter(Of LRMode)
    End Class

    Private Enum LRMode
        Stereo

        RightToBoth

        Side

        LeftToBoth

        Mono

        MidSideToXY
    End Enum
#End Region

#Region "Unbounded Database Types"
    Private NotInheritable Class UnboundedSapphireRoot
        Inherits Root(Of UnboundedSapphireRoot)

        Friend Property Sapphire As UnboundedSapphire
    End Class

    Private NotInheritable Class UnboundedSapphire
        Inherits FieldNode(Of UnboundedSapphire)

        Friend Property Sources As CollectionNode(Of Source)
    End Class
#End Region

#Region "Optional Fader Source"
    Private NotInheritable Class OptionalFaderSource
        Inherits FieldNode(Of OptionalFaderSource)

        <Element(IsOptional:=True)>
        Friend Property Fader As Fader

        <Element(Identifier:="DSP")>
        Friend Property Dsp As Dsp
    End Class
#End Region

#Region "Mixed Database Types"
    ' Subclassing Root means that the Children collection of this node will only contain the elements declared
    ' with properties, in this case a single node with the identifier Sapphire, which is also accessible through
    ' the property.
    Private NotInheritable Class MixedSapphireRoot
        Inherits Root(Of MixedSapphireRoot)

        Friend Property Sapphire As MixedSapphire
    End Class

    ' Subclassing DynamicFieldNode means that the Children collection of this node will contain *all* elements
    ' reported by the provider. Additionally, the node with the identifier Sources is also accessible through the
    ' property.
    Private NotInheritable Class MixedSapphire
        Inherits DynamicFieldNode(Of MixedSapphire)

        Friend Property Sources As CollectionNode(Of MixedSource)
    End Class

    ' Subclassing DynamicFieldNode means that the Children collection of this node will contain *all* elements
    ' reported by the provider. Additionally, the nodes Fader and Dsp are also accessible through their
    ' respective properties. The Fader and Dsp types themselves derive from FieldNode, so their Children
    ' collections will only contain the parameters declared as properties.
    Private NotInheritable Class MixedSource
        Inherits DynamicFieldNode(Of MixedSource)

        Friend Property Fader As Fader

        <Element(Identifier:="DSP")>
        Friend Property Dsp() As Dsp
    End Class
#End Region
End Class

File: Lawo.EmberPlusSharpTest/Model/TutorialTest.cs
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
// <copyright>Copyright 2012-2017 Lawo AG (http://www.lawo.com).</copyright>
// Distributed under the Boost Software License, Version 1.0.
// (See accompanying file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

namespace Lawo.EmberPlusSharp.Model
{
    // The following suppressions are necessary so that tested code snippets can be included in the documentation.
#pragma warning disable SA1123 // Do not place regions within elements
#pragma warning disable SA1124 // Do not use regions
#pragma warning disable SA1515 // Single-line comment must be preceded by blank line
    #region Using Declarations
    using System;
    using System.Linq;
    using System.Net.Sockets;
    using System.Threading.Tasks;

    using Microsoft.VisualStudio.TestTools.UnitTesting;
    using S101;
    using Threading.Tasks;
    #endregion

    /// <summary>Tests the <see cref="Consumer{T}"/>-based code shown in the tutorial.</summary>
    [TestClass]
    public class TutorialTest
    {
        /// <summary>Establishes a connection to a provider and creates a dynamic local copy of the database.</summary>
        [TestMethod]
        [TestCategory("Manual")]
        public void DynamicConnectTest() => MainTutorial();

        /// <summary>Iterates over the dynamic local database.</summary>
        [TestMethod]
        [TestCategory("Manual")]
        public void DynamicIterateTest()
        {
            #region Dynamic Iterate
            var cancelToken = new CancellationTokenSource().Token;
            AsyncPump.Run(
                async () =>
                {
                    using (var client = await ConnectAsync("localhost", 9000))
                    using (var consumer = await Consumer<MyRoot>.CreateAsync(client))
                    {
                        WriteChildren(consumer.Root, 0);
                    }
                }, cancelToken);
            #endregion
        }

        /// <summary>Demonstrates how to react to changes with the dynamic interface.</summary>
        [TestMethod]
        [TestCategory("Manual")]
        public void DynamicReactToChangesTest()
        {
            #region Dynamic React to Changes
            var cancelToken = new CancellationTokenSource().Token;
            AsyncPump.Run(
                async () =>
                {
                    using (var client = await ConnectAsync("localhost", 9000))
                    using (var consumer = await Consumer<MyRoot>.CreateAsync(client))
                    {
                        INode root = consumer.Root;

                        // Navigate to the parameter we're interested in.
                        var sapphire = (INode)root.Children.First(c => c.Identifier == "Sapphire");
                        var sources = (INode)sapphire.Children.First(c => c.Identifier == "Sources");
                        var fpgm1 = (INode)sources.Children.First(c => c.Identifier == "FPGM 1");
                        var fader = (INode)fpgm1.Children.First(c => c.Identifier == "Fader");
                        var positionParameter = fader.Children.First(c => c.Identifier == "Position");

                        var valueChanged = new TaskCompletionSource<string>();
                        positionParameter.PropertyChanged += (s, e) => valueChanged.SetResult(((IElement)s).GetPath());
                        Console.WriteLine("Waiting for the parameter to change...");
                        Console.WriteLine("A value of the element with the path {0} has been changed.", await valueChanged.Task);
                    }
                }, cancelToken);
            #endregion
        }

        /// <summary>Modifies parameters in the dynamic local database.</summary>
        [TestMethod]
        [TestCategory("Manual")]
        public void DynamicModifyTest()
        {
            #region Dynamic Modify
            var cancelToken = new CancellationTokenSource().Token;
            AsyncPump.Run(
                async () =>
                {
                    using (var client = await ConnectAsync("localhost", 9000))
                    using (var consumer = await Consumer<MyRoot>.CreateAsync(client))
                    {
                        INode root = consumer.Root;

                        // Navigate to the parameters we're interested in.
                        var sapphire = (INode)root.Children.First(c => c.Identifier == "Sapphire");
                        var sources = (INode)sapphire.Children.First(c => c.Identifier == "Sources");
                        var fpgm1 = (INode)sources.Children.First(c => c.Identifier == "FPGM 1");
                        var fader = (INode)fpgm1.Children.First(c => c.Identifier == "Fader");
                        var level = (IParameter)fader.Children.First(c => c.Identifier == "dB Value");
                        var position = (IParameter)fader.Children.First(c => c.Identifier == "Position");

                        // Set parameters to the desired values.
                        level.Value = -67.0;
                        position.Value = 128L;

                        // We send the changes back to the provider with the call below. Here, this is necessary so that
                        // the changes are sent before Dispose is called on the consumer. In a real-world application
                        // however, SendAsync often does not need to be called explicitly because it is automatically
                        // called every 100ms as long as there are pending changes. See AutoSendInterval for more
                        // information.
                        await consumer.SendAsync();
                    }
                }, cancelToken);
            #endregion
        }

        /// <summary>Waits for the connection to be lost.</summary>
        [TestMethod]
        [TestCategory("Manual")]
        public void ConnectionLostTest()
        {
            #region Connection Lost
            var cancelToken = new CancellationTokenSource().Token;
            AsyncPump.Run(
                async () =>
                {
                    using (var client = await ConnectAsync("localhost", 9000))
                    using (var consumer = await Consumer<MyRoot>.CreateAsync(client))
                    {
                        var connectionLost = new TaskCompletionSource<Exception>();
                        consumer.ConnectionLost += (s, e) => connectionLost.SetResult(e.Exception);

                        Console.WriteLine("Waiting for the provider to disconnect...");
                        var exception = await connectionLost.Task;
                        Console.WriteLine("Connection Lost!");
                        Console.WriteLine("Exception:{0}{1}", exception, Environment.NewLine);
                    }
                }, cancelToken);
            #endregion
        }

        /// <summary>Iterates over the static local database.</summary>
        [TestMethod]
        [TestCategory("Manual")]
        public void StaticIterateTest()
        {
            #region Static Iterate
            var cancelToken = new CancellationTokenSource().Token;
            AsyncPump.Run(
                async () =>
                {
                    using (var client = await ConnectAsync("localhost", 9000))
                    using (var consumer = await Consumer<SapphireRoot>.CreateAsync(client))
                    {
                        WriteChildren(consumer.Root, 0);
                    }
                }, cancelToken);
            #endregion
        }

        /// <summary>Demonstrates how to react to changes with the static interface.</summary>
        [TestMethod]
        [TestCategory("Manual")]
        public void StaticReactToChangesTest()
        {
            #region Static React to Changes
            var cancelToken = new CancellationTokenSource().Token;
            AsyncPump.Run(
                async () =>
                {
                    using (var client = await ConnectAsync("localhost", 9000))
                    using (var consumer = await Consumer<SapphireRoot>.CreateAsync(client))
                    {
                        var valueChanged = new TaskCompletionSource<string>();
                        var positionParameter = consumer.Root.Sapphire.Sources.Fpgm1.Fader.Position;
                        positionParameter.PropertyChanged += (s, e) => valueChanged.SetResult(((IElement)s).GetPath());
                        Console.WriteLine("Waiting for the parameter to change...");
                        Console.WriteLine("A value of the element with the path {0} has been changed.", await valueChanged.Task);
                    }
                }, cancelToken);
            #endregion
        }

        /// <summary>Modifies parameters in the dynamic local database.</summary>
        [TestMethod]
        [TestCategory("Manual")]
        public void StaticModifyTest()
        {
            #region Static Modify
            var cancelToken = new CancellationTokenSource().Token;
            AsyncPump.Run(
                async () =>
                {
                    using (var client = await ConnectAsync("localhost", 9000))
                    using (var consumer = await Consumer<SapphireRoot>.CreateAsync(client))
                    {
                        var fader = consumer.Root.Sapphire.Sources.Fpgm1.Fader;
                        fader.DBValue.Value = -67.0;
                        fader.Position.Value = 128;
                        await consumer.SendAsync();
                    }
                }, cancelToken);
            #endregion
        }

        /// <summary>Tests <see cref="CollectionNode{T}"/>.</summary>
        [TestMethod]
        [TestCategory("Manual")]
        public void CollectionNodeTest()
        {
            #region Collection Node
            var cancelToken = new CancellationTokenSource().Token;
            AsyncPump.Run(
                async () =>
                {
                    using (var client = await ConnectAsync("localhost", 9000))
                    using (var consumer = await Consumer<UnboundedSapphireRoot>.CreateAsync(client))
                    {
                        foreach (var source in consumer.Root.Sapphire.Sources.Children)
                        {
                            Console.WriteLine(source.Fader.Position.Value);
                        }
                    }
                }, cancelToken);
            #endregion
        }

        /// <summary>Iterates over the mixed local database.</summary>
        [TestMethod]
        [TestCategory("Manual")]
        public void MixedIterateTest()
        {
            #region Mixed Iterate
            var cancelToken = new CancellationTokenSource().Token;
            AsyncPump.Run(
                async () =>
                {
                    using (var client = await ConnectAsync("localhost", 9000))
                    using (var consumer = await Consumer<MixedSapphireRoot>.CreateAsync(client))
                    {
                        WriteChildren(consumer.Root, 0);
                    }
                }, cancelToken);
            #endregion
        }

        /// <summary>Modifies parameters in the mixed local database.</summary>
        [TestMethod]
        [TestCategory("Manual")]
        public void MixedModifyTest()
        {
            #region Mixed Modify
            var cancelToken = new CancellationTokenSource().Token;
            AsyncPump.Run(
                async () =>
                {
                    using (var client = await ConnectAsync("localhost", 9000))
                    using (var consumer = await Consumer<MixedSapphireRoot>.CreateAsync(client))
                    {
                        foreach (var source in consumer.Root.Sapphire.Sources.Children)
                        {
                            source.Fader.DBValue.Value = -67.0;
                            source.Fader.Position.Value = 128;
                            source.Dsp.Input.LRMode.Value = LRMode.Mono;
                            source.Dsp.Input.Phase.Value = false;
                        }

                        await consumer.SendAsync();
                    }
                }, cancelToken);
            #endregion
        }

        ////////////////////////////////////////////////////////////////////////////////////////////////////////////////

        #region MainTutorial Method
        private static void MainTutorial()
        {
            // This is necessary so that we can execute async code in a console application.
            var cancelToken = new CancellationTokenSource().Token;
            AsyncPump.Run(
                async () =>
                {
                    // Establish S101 protocol
                    using (S101Client client = await ConnectAsync("localhost", 9000))

                    // Retrieve *all* elements in the provider database and store them in a local copy
                    using (Consumer<MyRoot> consumer = await Consumer<MyRoot>.CreateAsync(client))
                    {
                        // Get the root of the local database.
                        INode root = consumer.Root;

                        // For now just output the number of direct children under the root node.
                        Console.WriteLine(root.Children.Count);
                    }
                }, cancelToken);
        }
        #endregion

        #region S101 Connect Method
        private static async Task<S101Client> ConnectAsync(string host, int port)
        {
            // Create TCP connection
            var tcpClient = new TcpClient();
            await tcpClient.ConnectAsync(host, port);

            // Establish S101 protocol
            // S101 provides message packaging, CRC integrity checks and a keep-alive mechanism.
            var stream = tcpClient.GetStream();
            return new S101Client(tcpClient, stream.ReadAsync, stream.WriteAsync);
        }
        #endregion

        #region Write Children
        private static void WriteChildren(INode node, int depth)
        {
            var indent = new string(' ', 2 * depth);

            foreach (var child in node.Children)
            {
                var childNode = child as INode;

                if (childNode != null)
                {
                    Console.WriteLine("{0}Node {1}", indent, child.Identifier);
                    WriteChildren(childNode, depth + 1);
                }
                else
                {
                    var childParameter = child as IParameter;

                    if (childParameter != null)
                    {
                        Console.WriteLine("{0}Parameter {1}: {2}", indent, child.Identifier, childParameter.Value);
                    }
                }
            }
        }
        #endregion

        #region Dynamic Root Class
        // Note that the most-derived subtype MyRoot needs to be passed to the generic base class.
        private sealed class MyRoot : DynamicRoot<MyRoot>
        {
        }
        #endregion

#pragma warning disable SA1201 // Elements must appear in the correct order. Enumeration must appear last to be consistent with order of other types.
        #region Static Database Types
        private sealed class SapphireRoot : Root<SapphireRoot>
        {
            internal Sapphire Sapphire { get; private set; }
        }

        private sealed class Sapphire : FieldNode<Sapphire>
        {
            internal Sources Sources { get; private set; }
        }

        private sealed class Sources : FieldNode<Sources>
        {
            [Element(Identifier = "FPGM 1")]
            internal Source Fpgm1 { get; private set; }

            [Element(Identifier = "FPGM 2")]
            internal Source Fpgm2 { get; private set; }
        }

        private sealed class Source : FieldNode<Source>
        {
            internal Fader Fader { get; private set; }

            [Element(Identifier = "DSP")]
            internal Dsp Dsp { get; private set; }
        }

        private sealed class Fader : FieldNode<Fader>
        {
            [Element(Identifier = "dB Value")]
            internal RealParameter DBValue { get; private set; }

            internal IntegerParameter Position { get; private set; }
        }

        private sealed class Dsp : FieldNode<Dsp>
        {
            internal Input Input { get; private set; }
        }

        private sealed class Input : FieldNode<Input>
        {
            internal BooleanParameter Phase { get; private set; }

            [Element(Identifier = "LR Mode")]
            internal EnumParameter<LRMode> LRMode { get; private set; }
        }

        private enum LRMode
        {
            Stereo,

            RightToBoth,

            Side,

            LeftToBoth,

            Mono,

            MidSideToXY
        }
        #endregion
#pragma warning restore SA1201 // Elements must appear in the correct order

        #region Unbounded Database Types
        private sealed class UnboundedSapphireRoot : Root<UnboundedSapphireRoot>
        {
            internal UnboundedSapphire Sapphire { get; private set; }
        }

        private sealed class UnboundedSapphire : FieldNode<UnboundedSapphire>
        {
            internal CollectionNode<Source> Sources { get; private set; }
        }
        #endregion

        #region Optional Fader Source
        private sealed class OptionalFaderSource : FieldNode<OptionalFaderSource>
        {
            [Element(IsOptional = true)]
            internal Fader Fader { get; private set; }

            [Element(Identifier = "DSP")]
            internal Dsp Dsp { get; private set; }
        }
        #endregion

        #region Mixed Database Types
        // Subclassing Root means that the Children collection of this node will only contain the elements declared
        // with properties, in this case a single node with the identifier Sapphire, which is also accessible through
        // the property.
        private sealed class MixedSapphireRoot : Root<MixedSapphireRoot>
        {
            internal MixedSapphire Sapphire { get; private set; }
        }

        // Subclassing DynamicFieldNode means that the Children collection of this node will contain *all* elements
        // reported by the provider. Additionally, the node with the identifier Sources is also accessible through the
        // property.
        private sealed class MixedSapphire : DynamicFieldNode<MixedSapphire>
        {
            internal CollectionNode<MixedSource> Sources { get; private set; }
        }

        // Subclassing DynamicFieldNode means that the Children collection of this node will contain *all* elements
        // reported by the provider. Additionally, the nodes Fader and Dsp are also accessible through their
        // respective properties. The Fader and Dsp types themselves derive from FieldNode, so their Children
        // collections will only contain the parameters declared as properties.
        private sealed class MixedSource : DynamicFieldNode<MixedSource>
        {
            internal Fader Fader { get; private set; }

            [Element(Identifier = "DSP")]
            internal Dsp Dsp { get; private set; }
        }
        #endregion
    }
#pragma warning restore SA1515 // Single-line comment must be preceded by blank line
#pragma warning restore SA1124 // Do not use regions
#pragma warning restore SA1123 // Do not place regions within elements
}
